File: backend/internal/services/audit/audit.go
// Package audit provides a lightweight helper for recording admin mutations.
// It writes a single row to the audit_logs table and never panics — failures
// are logged but do not block the caller.
package audit

import (
	"log/slog"

	"github.com/AlexArtaud-Dev/velar/backend/internal/database"
	"github.com/AlexArtaud-Dev/velar/backend/internal/models"
)

// Log writes one audit entry. adminID may be 0 for system-initiated actions.
// action should follow the "<target>.<verb>" convention, e.g. "client.create".
func Log(adminID uint, action, targetType string, targetID uint, targetName, detail string) {
	entry := models.AuditLog{
		AdminID:    adminID,
		Action:     action,
		TargetType: targetType,
		TargetID:   targetID,
		TargetName: targetName,
		Detail:     detail,
	}
	if err := database.DB.Create(&entry).Error; err != nil {
		slog.Warn("audit: failed to write log", "action", action, "err", err)
	}
}

File: frontend/src/api/audit.ts
import { api } from './client'

export interface AuditLog {
  id: number
  admin_id: number
  action: string
  target_type: string
  target_id: number
  target_name: string
  detail: string
  created_at: string
}

export interface AuditLogResponse {
  total: number
  page: number
  limit: number
  items: AuditLog[]
}

export const listAuditLogs = (params?: { page?: number; limit?: number; action?: string }) =>
  api.get<AuditLogResponse>('/audit', { params }).then((r) => r.data)
